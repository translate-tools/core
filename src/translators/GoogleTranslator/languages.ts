import { LanguageAliases } from '../../languages/LanguageAliases';

/**
 * Raw languages array
 */
// prettier-ignore
export const supportedLanguages = [
	'af', 'ak', 'am', 'ar', 'as', 'ay', 'az', 'be', 'bg', 'bho',
	'bm', 'bn', 'bs', 'ca', 'ceb', 'ckb', 'co', 'cs', 'cy', 'da',
	'de', 'doi', 'dv', 'ee', 'el', 'en', 'eo', 'es', 'et', 'eu',
	'fa', 'fi', 'fr', 'fy', 'ga', 'gd', 'gl', 'gn', 'gom', 'gu',
	'ha', 'haw', 'hi', 'hmn', 'hr', 'ht', 'hu', 'hy', 'id', 'ig',
	'ilo', 'is', 'it', 'iw', 'ja', 'jw', 'ka', 'kk', 'km', 'kn',
	'ko', 'kri', 'ku', 'ky', 'la', 'lb', 'lg', 'ln', 'lo', 'lt',
	'lus', 'lv', 'mai', 'mg', 'mi', 'mk', 'ml', 'mn', 'mni-Mtei', 'mr',
	'ms', 'mt', 'my', 'ne', 'nl', 'no', 'nso', 'ny', 'om', 'or',
	'pa', 'pl', 'ps', 'pt', 'qu', 'ro', 'ru', 'rw', 'sa', 'sd',
	'si', 'sk', 'sl', 'sm', 'sn', 'so', 'sq', 'sr', 'st', 'su',
	'sv', 'sw', 'ta', 'te', 'tg', 'th', 'ti', 'tk', 'tl', 'tr',
	'ts', 'tt', 'ug', 'uk', 'ur', 'uz', 'vi', 'xh', 'yi', 'yo',
	'zh', 'zh-CN', 'zh-TW', 'zu'
];

/**
 * Map with languages aliases.
 *
 * Google translator use legacy codes for some languages,
 * this map useful to use actual language codes by aliases
 *
 * @link https://xml.coverpages.org/iso639a.html
 */
export const fixedLanguagesMap: Record<string, string> = {
	he: 'iw',
	jv: 'jw',
};

/**
 * Map ISO lang codes to google translator lang codes
 */
export const languageAliases = new LanguageAliases(supportedLanguages, {
	map: fixedLanguagesMap,
});

/**
 * @param language language code or `auto`
 * @returns mapped language
 */
export const getFixedLanguage = (language: string) =>
	languageAliases.get(language) ?? language;
