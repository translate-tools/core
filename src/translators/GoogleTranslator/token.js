// This file is generated, so eslint is not needed here
/* eslint-disable */
// TKK value from https://github.com/FilipePS/Traduzir-paginas-web/blob/f3a4956a1aa96b7a9124864158a5200827694521/background/translationService.js
const googleTranslateTKK = '448487.932609646';

//
// Obfuscated logic: START
//

function shiftLeftOrRightThenSumOrXor(num, optString) {
	for (let i = 0; i < optString.length - 2; i += 3) {
		let acc = optString.charAt(i + 2);
		if ('a' <= acc) {
			acc = acc.charCodeAt(0) - 87;
		} else {
			acc = Number(acc);
		}
		if (optString.charAt(i + 1) == '+') {
			acc = num >>> acc;
		} else {
			acc = num << acc;
		}
		if (optString.charAt(i) == '+') {
			num += acc & 4294967295;
		} else {
			num ^= acc;
		}
	}
	return num;
}

function transformQuery(query) {
	const bytesArray = [];
	let idx = [];
	for (let i = 0; i < query.length; i++) {
		let charCode = query.charCodeAt(i);

		if (128 > charCode) {
			bytesArray[idx++] = charCode;
		} else {
			if (2048 > charCode) {
				bytesArray[idx++] = (charCode >> 6) | 192;
			} else {
				if (
					55296 == (charCode & 64512) &&
					i + 1 < query.length &&
					56320 == (query.charCodeAt(i + 1) & 64512)
				) {
					charCode =
						65536 +
						((charCode & 1023) << 10) +
						(query.charCodeAt(++i) & 1023);
					bytesArray[idx++] = (charCode >> 18) | 240;
					bytesArray[idx++] = ((charCode >> 12) & 63) | 128;
				} else {
					bytesArray[idx++] = (charCode >> 12) | 224;
				}
				bytesArray[idx++] = ((charCode >> 6) & 63) | 128;
			}
			bytesArray[idx++] = (charCode & 63) | 128;
		}
	}
	return bytesArray;
}

function calcHash(query, windowTkk) {
	const tkkSplited = windowTkk.split('.');
	const tkkIndex = Number(tkkSplited[0]) || 0;
	const tkkKey = Number(tkkSplited[1]) || 0;

	const bytesArray = transformQuery(query);

	let encondingRound = tkkIndex;
	for (let i = 0; i < bytesArray.length; i++) {
		encondingRound += bytesArray[i];
		encondingRound = shiftLeftOrRightThenSumOrXor(encondingRound, '+-a^+6');
	}
	encondingRound = shiftLeftOrRightThenSumOrXor(encondingRound, '+-3^+b+-f');

	encondingRound ^= tkkKey;
	if (encondingRound <= 0) {
		encondingRound = (encondingRound & 2147483647) + 2147483648;
	}

	const normalizedResult = encondingRound % 1000000;
	return normalizedResult.toString() + '.' + (normalizedResult ^ tkkIndex);
}

//
// Obfuscated logic: END
//

export function getToken(query) {
	return new Promise((res) => res({ value: calcHash(query, googleTranslateTKK) }));
}
