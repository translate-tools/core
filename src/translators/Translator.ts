/**
 * @deprecated this type will be removed in future, use `string` instead
 */
export type langCode = string;

/**
 * @deprecated this type will be removed in future, use `string` instead
 */
export type langCodeWithAuto = string;

export interface TranslatorInstanceMembers {
	/**
	 * Translate text
	 * @returns Translated string
	 */
	translate(
		text: string,
		sourceLanguage: string,
		targetLanguage: string,
	): Promise<string>;

	/**
	 * Translate texts array
	 * @returns Array with translated strings and same length as input array
	 * @returns Texts that did not translated will replaced to `null`
	 */
	translateBatch(
		text: string[],
		sourceLanguage: string,
		targetLanguage: string,
	): Promise<(string | null)[]>;

	/**
	 * Check string or array of stings to exceeding a limit
	 *
	 * It need for modules with complexity logic of encoding a strings.
	 * For example, when in `translateBatch` text merge to string
	 * and split to chunks by tokens with ID: `<id1>Text 1</id1><id2>Text 2</id2>`
	 *
	 * Here checked result of encoding a input data
	 * @returns number of extra chars
	 */
	checkLimitExceeding(text: string | string[]): number;

	/**
	 * Check supporting of translate direction
	 */
	checkDirection?: (sourceLanguage: string, targetLanguage: string) => boolean;

	/**
	 * Max length of string for `translate` or total length of strings from array for `translateBatch`
	 */
	getLengthLimit(): number;

	/**
	 * Delay between requests that required by translator to a correct work
	 */
	getRequestsTimeout: () => number;
}

/**
 * Static members useful to choose most suitable translator in list by features
 */
export interface TranslatorStaticMembers {
	/**
	 * Public translator name to displaying
	 */
	readonly translatorName: string;

	/**
	 * Is required API key for this module
	 */
	isRequiredKey(): boolean;

	/**
	 * Is it supported value `auto` in `sourceLanguage` argument of `translate` and `translateBatch` methods
	 */
	isSupportedAutoFrom(): boolean;

	/**
	 * Array of supported languages as ISO 639-1 codes
	 */
	getSupportedLanguages(): string[];
}

/**
 * Translator class contract
 *
 * This type describe static members and members for instance both
 */
export interface TranslatorConstructor<
	InstanceProps extends TranslatorInstanceMembers = TranslatorInstanceMembers,
> extends TranslatorStaticMembers {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	new (...args: any[]): InstanceProps;
}
