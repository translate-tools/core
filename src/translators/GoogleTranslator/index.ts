import { stringify } from 'query-string';
import xpath from 'xpath';
import { DOMParser } from '@xmldom/xmldom';

import { langCode, langCodeWithAuto } from '../Translator';
import { BaseTranslator } from '../BaseTranslator';
import { getToken } from './token';
import { visitArrayItems } from './utils';

/**
 * Map with languages aliases.
 *
 * Google translator use legacy codes for some languages,
 * this map useful to use actual language codes by aliases
 *
 * @link https://xml.coverpages.org/iso639a.html
 */
const fixedLanguagesMap: Record<string, string> = {
	he: 'iw',
	jv: 'jw',
};

/**
 * Raw languages array
 */
// eslint-disable
// prettier-ignore
export const supportedLanguages = [
	'af', 'ak', 'am', 'ar', 'as', 'ay', 'az', 'be', 'bg', 'bho',
	'bm', 'bn', 'bs', 'ca', 'ceb', 'ckb', 'co', 'cs', 'cy', 'da',
	'de', 'doi', 'dv', 'ee', 'el', 'en', 'eo', 'es', 'et', 'eu',
	'fa', 'fi', 'fr', 'fy', 'ga', 'gd', 'gl', 'gn', 'gom', 'gu',
	'ha', 'haw', 'hi', 'hmn', 'hr', 'ht', 'hu', 'hy', 'id', 'ig',
	'ilo', 'is', 'it', 'iw', 'ja', 'jw', 'ka', 'kk', 'km', 'kn',
	'ko', 'kri', 'ku', 'ky', 'la', 'lb', 'lg', 'ln', 'lo', 'lt',
	'lus', 'lv', 'mai', 'mg', 'mi', 'mk', 'ml', 'mn', 'mni-Mtei', 'mr',
	'ms', 'mt', 'my', 'ne', 'nl', 'no', 'nso', 'ny', 'om', 'or',
	'pa', 'pl', 'ps', 'pt', 'qu', 'ro', 'ru', 'rw', 'sa', 'sd',
	'si', 'sk', 'sl', 'sm', 'sn', 'so', 'sq', 'sr', 'st', 'su',
	'sv', 'sw', 'ta', 'te', 'tg', 'th', 'ti', 'tk', 'tl', 'tr',
	'ts', 'tt', 'ug', 'uk', 'ur', 'uz', 'vi', 'xh', 'yi', 'yo',
	'zh', 'zh-CN', 'zh-TW', 'zu'
];
// eslint-enable

const parseXMLResponse = (text: string) => {
	let doc: Document;
	try {
		doc = new DOMParser().parseFromString(text);
	} catch (err) {
		console.error(err);
		return null;
	}

	const nodesWithTranslation = xpath.select('//pre/*[not(self::i)]', doc);

	if (nodesWithTranslation.length === 0) return null;
	// console.log('Selected nodes', nodesWithTranslation.map((node) => (node as Node).toString()));
	return nodesWithTranslation
		.map((node) => {
			// Select text in child nodes or in self
			const textNodes = xpath.select('descendant-or-self::*/text()', node as Node);
			if (textNodes.length > 1) {
				console.debug('More than one text node found');
			}

			return textNodes.length === 0 ? '' : textNodes.join(' ');
		})
		.join(' ');
};

/**
 * Common class for google translator implementations
 */
export abstract class AbstractGoogleTranslator extends BaseTranslator {
	public static isSupportedAutoFrom() {
		return true;
	}

	public static getSupportedLanguages(): langCode[] {
		return supportedLanguages;
	}

	public getLengthLimit() {
		return 4000;
	}

	public getRequestsTimeout() {
		return 300;
	}

	/**
	 * Map ISO lang codes to google translator lang codes
	 */
	protected getFixedLanguage(lang: langCodeWithAuto) {
		if (lang === 'zh') return 'zh-CN';
		return fixedLanguagesMap[lang] ?? lang;
	}
}

/**
 * Translator implementation which use Google API with token from https://translate.google.com
 */
export class GoogleTranslator extends AbstractGoogleTranslator {
	public static readonly translatorName = 'GoogleTranslator';

	public checkLimitExceeding(text: string | string[]) {
		if (Array.isArray(text)) {
			const encodedText = this.encodeForBatch(text).join('');
			const extra = encodedText.length - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		} else {
			const extra = text.length - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		}
	}

	public translate(text: string, from: langCodeWithAuto, to: langCode) {
		return getToken(text).then(({ value: tk }) => {
			const apiPath = 'https://translate.google.com/translate_a/single';

			const data = {
				client: 't',
				sl: this.getFixedLanguage(from),
				tl: this.getFixedLanguage(to),
				hl: this.getFixedLanguage(to),
				dt: ['at', 'bd', 'ex', 'ld', 'md', 'qca', 'rw', 'rm', 'ss', 't'],
				ie: 'UTF-8',
				oe: 'UTF-8',
				otf: 1,
				ssel: 0,
				tsel: 0,
				kc: 7,
				q: text,
				tk,
			};

			const url = apiPath + '?' + stringify(data);

			return this.fetch(this.wrapUrlToCorsProxy(url), {
				responseType: 'json',
				method: 'GET',
				headers: this.options.headers,
			})
				.then((rsp) => rsp.data)
				.then((rsp) => {
					if (!(rsp instanceof Array) || !(rsp[0] instanceof Array)) {
						throw new Error('Unexpected response');
					}

					const translatedText = rsp[0]
						.map((chunk) =>
							chunk instanceof Array && typeof chunk[0] === 'string'
								? chunk[0]
								: '',
						)
						.join('');

					return translatedText;
				});
		});
	}

	public translateBatch(text: string[], from: langCodeWithAuto, to: langCode) {
		const preparedText = this.encodeForBatch(text);
		return getToken(preparedText.join('')).then(({ value: tk }) => {
			const apiPath = 'https://translate.googleapis.com/translate_a/t';

			const data = {
				anno: 3,
				client: 'te',
				v: '1.0',
				format: 'html',
				sl: this.getFixedLanguage(from),
				tl: this.getFixedLanguage(to),
				tk,
			};

			const url = apiPath + '?' + stringify(data);
			const body = preparedText
				.map((text) => `&q=${encodeURIComponent(text)}`)
				.join('');

			return this.fetch(this.wrapUrlToCorsProxy(url), {
				responseType: 'json',
				method: 'POST',
				headers: {
					'Content-Type': 'application/x-www-form-urlencoded',
					...this.options.headers,
				},
				body,
			})
				.then((rsp) => rsp.data)
				.then((rawResp) => {
					try {
						if (!Array.isArray(rawResp)) {
							throw new Error('Unexpected response');
						}

						const isSingleResponseMode = text.length === 1;

						const result: string[] = [];
						visitArrayItems(rawResp, (obj) => {
							if (isSingleResponseMode && result.length === 1) return;

							if (typeof obj !== 'string') return;

							if (isSingleResponseMode) {
								const parsedText = parseXMLResponse(obj);
								result.push(parsedText || obj);
							} else {
								const parsedText = parseXMLResponse(obj);
								if (parsedText !== null) {
									result.push(parsedText);
								}
							}
						});

						if (result.length !== text.length) {
							throw new Error(
								'Mismatching a lengths of original and translated arrays',
							);
						}

						return result as string[];
					} catch (err) {
						console.warn('Got response', rawResp);
						throw err;
					}
				});
		});
	}

	private encodeForBatch(textList: string[]) {
		return textList.map((text, i) => `<pre><a i="${i}">${text}</a></pre>`);
	}
}

/**
 * Translator implementation which use Google API without token
 */
export class GoogleTranslatorTokenFree extends AbstractGoogleTranslator {
	public static readonly translatorName = 'GoogleTranslatorTokenFree';

	public translate = async (text: string, from: langCodeWithAuto, to: langCode) => {
		const [translation] = await this.translateBatch([text], from, to);
		return translation;
	};

	public translateBatch(text: string[], from: langCodeWithAuto, to: langCode) {
		const apiPath = 'https://translate.googleapis.com/translate_a/t';

		const data = {
			client: 'dict-chrome-ex',
			sl: this.getFixedLanguage(from),
			tl: this.getFixedLanguage(to),
			q: text,
		};

		const url = apiPath + '?' + stringify(data);

		return this.fetch(this.wrapUrlToCorsProxy(url), {
			responseType: 'json',
			method: 'GET',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
				...this.options.headers,
			},
		})
			.then((rsp) => rsp.data)
			.then((rawResp) => {
				try {
					if (!Array.isArray(rawResp)) {
						throw new Error('Unexpected response');
					}

					const intermediateTextsArray: string[] = [];
					visitArrayItems(rawResp, (obj) => {
						if (typeof obj === 'string') {
							intermediateTextsArray.push(obj);
						}
					});

					const result: string[] = [];

					const isSingleResponseMode = text.length === 1;
					const isOneToOneMappingMode =
						intermediateTextsArray.length === text.length;
					for (const idx in intermediateTextsArray) {
						const text = intermediateTextsArray[idx];

						if (isSingleResponseMode) {
							result.push(text);
							break;
						}

						// Each second text it's not translation if not 1-1 mapping
						const isTranslation =
							isOneToOneMappingMode || Number(idx) % 2 === 0;
						if (isTranslation) {
							result.push(text);
						}
					}

					if (result.length !== text.length) {
						console.warn('Translation result', result);
						throw new Error(
							'Mismatching a lengths of original and translated arrays',
						);
					}

					return result as string[];
				} catch (err) {
					console.warn('Got response', rawResp);
					throw err;
				}
			});
	}
}
