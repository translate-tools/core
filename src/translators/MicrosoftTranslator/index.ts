import { langCode, langCodeWithAuto } from '../Translator';
import { BaseTranslator } from '../BaseTranslator';
import { z } from 'zod';
import { LanguageAliases } from '../../languages/LanguageAliases';

const ResponseScheme = z
	.object({
		translations: z
			.object({
				text: z.string(),
			})
			.array(),
	})
	.array()
	.or(
		z.object({
			error: z.object({
				code: z.number(),
				message: z.string(),
			}),
		}),
	);

// eslint-disable
// prettier-ignore
const languagesMap = new LanguageAliases([
	"ace",
	"af",
	"sq",
	"am",
	"ar",
	"arz",
	"ary",
	"arb",
	"hy",
	"as",
	"ast",
	"az",
	"ban",
	"bn",
	"ba",
	"eu",
	"bbc",
	"be",
	"bho",
	"bik",
	"brx",
	"bs",
	"bg",
	"yue",
	"ca",
	"ceb",
	"hne",
	"lzh",
	"zh-Hans",
	"zh-Hant",
	"co",
	"hr",
	"cs",
	"da",
	"prs",
	"dv",
	"doi",
	"nl",
	"en",
	"en-GB",
	"epo",
	"et",
	"fo",
	"fj",
	"fil",
	"fi",
	"fr",
	"fr-CA",
	"fy",
	"fur",
	"gl",
	"lug",
	"ka",
	"de",
	"el",
	"gu",
	"ht",
	"ha",
	"he",
	"hil",
	"hi",
	"mww",
	"hu",
	"iba",
	"is",
	"ig",
	"ilo",
	"id",
	"ikt",
	"iu",
	"iu-Latn",
	"ga",
	"it",
	"jam",
	"ja",
	"jav",
	"kea",
	"kn",
	"pam",
	"ks",
	"kk",
	"km",
	"rw",
	"tlh-Latn",
	"gom",
	"ko",
	"kri",
	"ku",
	"kmr",
	"ky",
	"lo",
	"la",
	"lv",
	"lij",
	"lim",
	"ln",
	"lt",
	"lmo",
	"dsb",
	"lb",
	"mk",
	"mai",
	"mg",
	"ms",
	"ml",
	"mt",
	"mr",
	"mwr",
	"mfe",
	"min",
	"mn-Cyrl",
	"mn-Mong",
	"my",
	"mi",
	"ne",
	"nb",
	"nno",
	"nya",
	"oc",
	"or",
	"pap",
	"ps",
	"fa",
	"pl",
	"pt",
	"pt-PT",
	"pa",
	"pnb",
	"otq",
	"ro",
	"run",
	"ru",
	"sm",
	"sa",
	"srd",
	"sr-Cyrl",
	"sr-Latn",
	"st",
	"nso",
	"tn",
	"crs",
	"sn",
	"scn",
	"sd",
	"si",
	"sk",
	"sl",
	"so",
	"es",
	"su",
	"sw",
	"sv",
	"ty",
	"tgk",
	"ta",
	"tt",
	"te",
	"tet",
	"th",
	"bo",
	"ti",
	"tpi",
	"to",
	"tr",
	"tk",
	"uk",
	"hsb",
	"ur",
	"ug",
	"uz",
	"vec",
	"vi",
	"war",
	"cy",
	"xh",
	"ydd",
	"yo",
	"yua",
	"zu"
]);
// eslint-enable

const fetchToken = () =>
	fetch('https://edge.microsoft.com/translate/auth', {
		headers: {
			accept: '*/*',
			'accept-language': 'zh-TW,zh;q=0.9,ja;q=0.8,zh-CN;q=0.7,en-US;q=0.6,en;q=0.5',
			'cache-control': 'no-cache',
			pragma: 'no-cache',
			priority: 'u=1, i',
			'referrer-policy': 'strict-origin-when-cross-origin',
			'sec-fetch-dest': 'empty',
			'sec-fetch-mode': 'cors',
			'sec-fetch-site': 'none',
			'sec-fetch-storage-access': 'active',
		},
		referrerPolicy: 'strict-origin-when-cross-origin',
		body: null,
		method: 'GET',
		mode: 'cors',
		credentials: 'omit',
	}).then((r) => r.text());

export class MicrosoftTranslator extends BaseTranslator {
	public static readonly translatorName = 'MicrosoftTranslator';

	public static isSupportedAutoFrom = () => true;

	public static getSupportedLanguages(): langCode[] {
		return languagesMap.getAll();
	}

	public getLengthLimit() {
		return 5000;
	}

	public getRequestsTimeout() {
		return 300;
	}

	public checkLimitExceeding(text: string | string[]) {
		if (Array.isArray(text)) {
			const arrayLen = text.reduce((acc, text) => acc + text.length, 0);
			const extra = arrayLen - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		} else {
			const extra = text.length - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		}
	}

	public async translate(text: string, from: langCodeWithAuto, to: langCode) {
		return this.translateBatch([text], from, to).then((resp) => resp[0]);
	}

	private token:
		| {
				value: string;
				issuedAt: number;
		  }
		| Promise<string>
		| null = null;
	private async getToken() {
		// Wait resolution if pending
		if (this.token instanceof Promise) return this.token;

		// Fetch new token
		if (!this.token || this.token.issuedAt < 60_000) {
			this.token = fetchToken().then((token) => {
				this.token = { value: token, issuedAt: Date.now() };
				return token;
			});

			return await this.token;
		}

		// Use cached value
		return this.token.value;
	}

	public async translateBatch(text: string[], from: langCodeWithAuto, to: langCode) {
		const sourceLanguage = from === 'auto' ? 'auto' : languagesMap.get(from);
		const targetLanguage = languagesMap.get(to);

		if (!sourceLanguage) throw new TypeError(`Unsupported source language ${from}`);
		if (!targetLanguage) throw new TypeError(`Unsupported source language ${to}`);

		const token = await this.getToken();

		const url =
			'https://api-edge.cognitive.microsofttranslator.com/translate?' +
			// Omit `from` parameter for auto detection of language
			(sourceLanguage !== 'auto'
				? `from=${encodeURIComponent(sourceLanguage)}&`
				: '') +
			`to=${encodeURIComponent(
				targetLanguage,
			)}&api-version=3.0&includeSentenceLength=true`;

		return fetch(url, {
			headers: {
				accept: '*/*',
				'accept-language':
					'zh-TW,zh;q=0.9,ja;q=0.8,zh-CN;q=0.7,en-US;q=0.6,en;q=0.5',
				authorization: `Bearer ${token}`,
				'cache-control': 'no-cache',
				'content-type': 'application/json',
				pragma: 'no-cache',
				priority: 'u=1, i',
				'referrer-policy': 'strict-origin-when-cross-origin',
				'sec-fetch-dest': 'empty',
				'sec-fetch-mode': 'cors',
				'sec-fetch-site': 'none',
				'sec-fetch-storage-access': 'active',
			},
			referrerPolicy: 'strict-origin-when-cross-origin',
			body: JSON.stringify(text.map((text) => ({ Text: text }))),
			method: 'POST',
			mode: 'cors',
			credentials: 'include',
		})
			.then((r) => r.json())
			.then((rawResult) => {
				const result = ResponseScheme.parse(rawResult);

				if ('error' in result) {
					throw new Error(`Code ${result.error.code}: ${result.error.message}`);
				}

				// Transform translations array
				return result.map((translationItem) =>
					// Build translation for single text
					translationItem.translations
						.map((translationSegment) => translationSegment.text)
						.join(' '),
				);
			});
	}
}
