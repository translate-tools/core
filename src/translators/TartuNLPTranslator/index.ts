import { langCode, langCodeWithAuto } from '../Translator';
import { BaseTranslator } from '../BaseTranslator';

export class TartuNLPTranslator extends BaseTranslator {
	public static readonly translatorName = 'TartuNLPTranslator';

	public static isSupportedAutoFrom() {
		return false;
	}

	public static getSupportedLanguages(): langCode[] {
		// eslint-disable
		// prettier-ignore
		return [
			"en", "et", "de", "lt", "lv",
			"fi", "ru", "no", "hu", "se",
		];
		// eslint-enable
	}

	public getLengthLimit() {
		return 5000;
	}

	public getRequestsTimeout() {
		return 300;
	}

	public checkLimitExceeding(text: string | string[]) {
		if (Array.isArray(text)) {
			const arrayLen = text.reduce((acc, text) => acc + text.length, 0);
			const extra = arrayLen - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		} else {
			const extra = text.length - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		}
	}

	public async translate(text: string, from: langCodeWithAuto, to: langCode) {
		return this.translateBatch([text], from, to).then((resp) => resp[0]);
	}

	public async translateBatch(text: string[], from: langCodeWithAuto, to: langCode) {
		return this.fetch('https://api.tartunlp.ai/translation/v2', {
			responseType: 'json',
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
				...this.options.headers,
			},
			body: JSON.stringify({ text, src: from, tgt: to }),
		}).then((rsp) => {
			if (
				typeof rsp.data !== 'object' ||
				rsp.data === null ||
				!Array.isArray((rsp.data as any).result)
			) {
				throw new TypeError('Unexpected data');
			}

			return (rsp.data as any).result as string[];
		});
	}
}
