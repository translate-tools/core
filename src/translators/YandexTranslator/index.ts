import { unescape } from 'lodash';
import queryString from 'query-string';
import z from 'zod';

import { BaseTranslator } from '../BaseTranslator';
import { langCode, langCodeWithAuto } from '../Translator';
import { getYandexSID } from './getYandexSID';

export class YandexTranslator extends BaseTranslator {
	public static readonly translatorName = 'YandexTranslator';

	public static isSupportedAutoFrom() {
		return true;
	}

	public static getSupportedLanguages(): langCode[] {
		// Supported, but not valid languages ['mhr', 'pap', 'ceb', 'mrj', 'udm']

		// eslint-disable
		// prettier-ignore
		return [
			'af', 'am', 'ar', 'az', 'ba', 'be', 'bg', 'bn', 'bs', 'ca',
			'ceb', 'cs', 'cv', 'cy', 'da', 'de', 'el', 'emj', 'en', 'eo',
			'es', 'et', 'eu', 'fa', 'fi', 'fr', 'ga', 'gd', 'gl', 'gu',
			'he', 'hi', 'hr', 'ht', 'hu', 'hy', 'id', 'is', 'it', 'ja',
			'jv', 'ka', 'kazlat', 'kk', 'km', 'kn', 'ko', 'ky', 'la', 'lb',
			'lo', 'lt', 'lv', 'mg', 'mhr', 'mi', 'mk', 'ml', 'mn', 'mr',
			'mrj', 'ms', 'mt', 'my', 'ne', 'nl', 'no', 'pa', 'pap', 'pl',
			'pt', 'ro', 'ru', 'sah', 'si', 'sjn', 'sk', 'sl', 'sq', 'sr',
			'su', 'sv', 'sw', 'ta', 'te', 'tg', 'th', 'tl', 'tr', 'tt',
			'udm', 'uk', 'ur', 'uz', 'uzbcyr', 'vi', 'xh', 'yi', 'zh', 'zu'
		];
		// eslint-enable
	}

	public getLengthLimit() {
		return 4000;
	}

	public getRequestsTimeout() {
		return 500;
	}

	public checkLimitExceeding(text: string | string[]) {
		if (Array.isArray(text)) {
			const arrayLen = text.reduce((acc, text) => acc + text.length, 0);
			const extra = arrayLen - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		} else {
			const extra = text.length - this.getLengthLimit();
			return extra > 0 ? extra : 0;
		}
	}

	public async translate(text: string, from: langCodeWithAuto, to: langCode) {
		return this.translateBatch([text], from, to).then((resp) => resp[0]);
	}

	public async translateBatch(text: string[], from: langCodeWithAuto, to: langCode) {
		const sid = await getYandexSID(this.fetch);
		if (sid === null) {
			throw new Error('Invalid SID');
		}

		const options = {
			format: 'html',
			lang: from === 'auto' ? to : `${from}-${to}`,
		};

		let body = queryString.stringify(options);
		for (const textChunk of text) {
			body += '&text=' + encodeURIComponent(textChunk);
		}

		// NOTE: if service will resist and will not work, it may check order of headers, parameters and requests
		// in this case just make requests less specific to it looks like requests from typical page (with overhead requests if require)
		const urlWithSid =
			'https://translate.yandex.net/api/v1/tr.json/translate?srv=tr-url-widget&id=' +
			sid +
			'-0-0&';
		return this.fetch(urlWithSid + body, {
			responseType: 'json',
			method: 'GET',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
				...this.options.headers,
			},
		}).then((response) => {
			const data = z
				.object({
					text: z.string().array(),
				})
				.parse(response.data);

			return data.text.map((text) => unescape(text));
		});
	}
}
