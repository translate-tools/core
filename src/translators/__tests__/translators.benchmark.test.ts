import { getTranslatorsScore } from './benchmark';
import { TranslatorConstructor } from '../Translator';

import { GoogleTranslator, GoogleTranslatorTokenFree } from '../GoogleTranslator';
import { YandexTranslator } from '../YandexTranslator';
import { TartuNLPTranslator } from '../TartuNLPTranslator';
import { MicrosoftTranslator } from '../MicrosoftTranslator';

const referenceText = `
Linguist is a powerful browser extension that is ready to replace your favorite translation service.

Translate web pages, highlighted text, Netflix subtitles, private messages, speak the translated text, and save important translations to your personal dictionary to learn new words in 130 languages.

Why Linguist?

Unlike other browser extensions, Linguist is not just a wrapper over the Google Translator Widget; it's a full-featured and independent translation system. This is why with Linguist you can be private and translate texts offline on your device and use any translation service, even your own like ChatGPT. See a custom translators list to find bindings for the most popular translation services.

Linguist is a free, open-source project, respects your privacy, and does not collect your personal data.
`.trim();

const referenceTranslation = `
Linguist - это мощное расширение для браузера, готовое заменить ваш любимый сервис перевода.

Переводите веб-страницы, выделенный текст, субтитры Netflix, приватные сообщения, озвучивайте переведённый текст и сохраняйте важные переводы в словарь, чтобы изучать новые слова на 130 языках.

Почему Linguist?

В отличие от других расширений, Linguist не просто обёртка над Google Translator Widget; это полнофункциональная и независимая система перевода. Поэтому с Linguist вы можете сохранять конфиденциальность, переводя тексты офлайн прямо на своём устройстве и использовать любой сервис перевода, даже собственный, как например ChatGPT. Ознакомьтесь со списком пользовательских переводчиков, чтобы найти адаптеры к самым популярным сервисам.

Linguist бесплатный проект с открытым исходным кодом, уважает вашу конфиденциальность и не собирает личные данные.
`.trim();

const translators: TranslatorConstructor[] = [
	GoogleTranslator,
	GoogleTranslatorTokenFree,
	YandexTranslator,
	TartuNLPTranslator,
	MicrosoftTranslator,
];

test(
	'Top translators list with score',
	async () => {
		await expect(
			getTranslatorsScore(translators, {
				text: referenceText,
				translation: referenceTranslation,
			}).then((results) => {
				results.forEach((result) => {
					result.score = parseFloat(result.score.toFixed(2));
				});

				return results;
			}),
		).resolves.matchSnapshot();
	},
	{ timeout: translators.length * 10_000 },
);
