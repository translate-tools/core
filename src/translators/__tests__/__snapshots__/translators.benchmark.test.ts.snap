// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Top translators list with score 1`] = `
[
  {
    "name": "MicrosoftTranslator",
    "score": 0.98,
  },
  {
    "name": "YandexTranslator",
    "score": 0.98,
  },
  {
    "name": "TartuNLPTranslator",
    "score": 0.97,
  },
  {
    "name": "GoogleTranslator",
    "score": 0.93,
  },
  {
    "name": "GoogleTranslatorTokenFree",
    "score": 0.93,
  },
]
`;
